python/shotgun_model/shotgun_hierarchy_model.py

```diff
@@ -0,0 +1,855 @@
+    key data methods for setting up queries, customizing etc. Then you connect
+    your class to a :class:`~PySide.QtGui.QAbstractItemView` of some sort which
+    will display the result. 
+
+    The model stores a single column, lazy-loaded Shotgun Hierarchy as queried
+    via the ``nav_expand()`` python-api method. The structure of items in the
+    hierarchy mimics what is found in Shotgun as configured in each project's
+    *Tracking Settings*.
+    """
+
+    # data field that uniquely identifies an entity
+    SG_DATA_UNIQUE_ID_FIELD = "url"
+
+    # data role used to track whether more data has been fetched for items
+    SG_ITEM_FETCHED_MORE = QtCore.Qt.UserRole + 3
+
+    # use hierarchy items when building the model
+    SG_QUERY_MODEL_ITEM_CLASS = ShotgunHierarchyItem
+
+    def __init__(self, parent, schema_generation=0, bg_task_manager=None):
+        """
+        Initialize the Hierarcy model.
+
+        :param parent: The model's parent.
+        :type parent: :class:`~PySide.QtGui.QObject`
+
+        """
+
+        super(ShotgunHierarchyModel, self).__init__(parent, bg_task_manager)
+
+        self._schema_generation = schema_generation
+
+        # flag to indicate a full refresh
+        self._request_full_refresh = False
+
+        # is the model set up with a query?
+        self._has_query = False
+
+        # keeps track of the currently running queries by mapping the id
+        # returned by the data retriever to the path/url being queried
+        self._running_query_lookup = {}
+
+        # keep these icons around so they're not constantly being created
+        self._folder_icon = QtGui.QIcon(
+            ":tk-framework-shotgunutils/icon_Folder.png")
+        self._none_icon = QtGui.QIcon(
+            ":tk-framework-shotgunutils/icon_None_dark.png")
+
+        # Define the foreground color of "empty" items.
+        # These special items are used as placeholders in the tree where the
+        # parent has no children. An example would be `Shots > No Shots` where
+        # `No Shots` is the "empty" item. By default, the color is a mix of the
+        # application instance's base and text colors. This will typically
+        # result in a dimmed appearance for these special items indicating that
+        # they are not clickable. This makes goes outside the typical bounds of
+        # the model by pulling the palette colors from the app instance. This
+        # can be overridden in subclasses via ``_finalize_item()`` though.
+        base_color = QtGui.QApplication.instance().palette().base().color()
+        text_color = QtGui.QApplication.instance().palette().text().color()
+        self._empty_item_color = color_mix(text_color, 1, base_color, 2)
+
+    ############################################################################
+    # public methods
+
+    def clear(self):
+        """
+        Removes all items (including header items) from the model and
```

Comment:
I think it would be great if we could point out which of these methods are subclassing QT4 methods and which ones are our own!